### PR TITLE
added Fn::Length and reworked Fn::Map to include index

### PR DESCRIPTION
**Pull request by arjanvandervelde**
Hi! As I needed to pull out index and length of sequences for replacement, I've added {placeholder}_IDX and {placeholder}_LEN as placeholders. I also added support for '!Condition' references and support for !GetAZs '' syntax.

In order to make it run in reasonable time, I unwound some of the recursiveness in `findAndReplace()`.

-- Thanks!

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -27,3 +27,4 @@ build/Release
 # Dependency directory
 # https://www.npmjs.org/doc/misc/npm-faq.html#should-i-check-my-node_modules-folder-into-git
 node_modules
+package-lock.json
```
Comment (monken) [inline .gitignore:30]: please use yarn for installation of dependencies. There is already a yarn.lock file.

File: index.js [hunk 2]
```diff
@@ -29,18 +28,24 @@ module.exports = function (options) {
...
 async function recurse(base, scope, object) {
+  let i = 0;
   scope = _.clone(scope);
   if (_.isArray(object)) return Promise.all(object.map(_.bind(recurse, this, base, scope)));
   else if (_.isPlainObject(object)) {
     if (object["Fn::Map"]) {
       var args = object["Fn::Map"],
         list = args[0],
         placeholder = args[1],
+        idx = `${placeholder}_IDX`,
```
Comment (monken) [inline index.js:39]: I'm not a fan of special variable names. Instead I'd prefer if the user would specify them:

```
Fn::Map:
  - [1, 2, 3]
  - [Number, Idx]
  - Resource{Number}:
      Type: {Idx}
```

File: index.js [hunk 3]
```diff
@@ -29,18 +28,24 @@ module.exports = function (options) {
...
+        len = `${placeholder}_LEN`,
```
Comment (monken) [inline index.js:40]: LEN should be it's own function, i.e. `Fn::Length` to stay consistent with popular map() functions (https://devdocs.io/javascript/global_objects/array/map)

File: index.js [hunk 4]
```diff
@@ -29,18 +28,24 @@ module.exports = function (options) {
 }
 
 async function recurse(base, scope, object) {
+  let i = 0;
```
Comment (monken) [inline index.js:31]: please add tests for this new feature.

**Review by monken — COMMENTED** [review]
Thank you very much for your contribution, please see comments attached. Also, please squash all changes into a single commit.

**Comment by arjanvandervelde** [comment]
Hi! Ok, I've added Fn::Length and reworked Fn::Map, added test cases and some documentation.

**Comment by monken** [comment]
merged

**Comment by arjanvandervelde** [comment]
thanks!

**Pull request closed without merging**